File: src/main/java/model/moves/MenCaptureBackwards.java
package model.moves;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import model.Cell;
import model.Checkers;
import model.Player;
import model.State;

public class MenCaptureBackwards extends Moves {

    public MenCaptureBackwards(Checkers c) {
        this.checkers = c;
    }

    @Override
    public List<List<Cell>> possibleMoves(Cell currentCell) {
        return this.possibleMoves(currentCell, new ArrayList<>());
    }

    private List<List<Cell>> possibleMoves(Cell currentCell, List<Cell> visitedCells) {
        State adversaryPlayer = (checkers.getTurn() == Player.PLAYER1) ? State.PLAYER2 : State.PLAYER1;
        int crow = currentCell.x(), ccol = currentCell.y();
        int[][] pos = {{-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
        List<List<Cell>> coords = new ArrayList<>();
        for (int[] po : pos) {
            int diffRow = po[0], diffCol = po[1];
            Cell destinationCell = new Cell(crow + 2 * diffRow, ccol + 2 * diffCol);
            Cell adversaryCell = new Cell(crow + diffRow, ccol + diffCol);
            if (!visitedCells.contains(adversaryCell) && checkers.onBoard(destinationCell) && checkers.getState(destinationCell) == State.EMPTY && checkers.getState(adversaryCell) == adversaryPlayer) {
                visitedCells.add(adversaryCell);
                List<List<Cell>> moves = possibleMoves(destinationCell, visitedCells);
                if (moves.isEmpty()) {
                    List<Cell> temp = new ArrayList<>();
                    temp.add(destinationCell);
                    coords.add(temp);
                } else {
                    moves.forEach(v -> v.add(destinationCell));
                    coords.addAll(moves);
                }
                visitedCells.remove(adversaryCell);
            }
        }
        int max = coords.stream().mapToInt(p -> p.size()).max().orElse(0);
        return coords.stream().filter(p -> p.size() == max).collect(Collectors.toList());
    }
}

File: docs/moves/MenCaptureBackwards.js
import Player from '../Player.js';
import Cell from '../Cell.js';
import State from '../State.js';

export default class MenCaptureBackwards {
    constructor(c) {
        this.checkers = c;
    }
    possibleMoves(currentCell, visitedCells = []) {
        let adversaryPlayer = (this.checkers.turn === Player.PLAYER1) ? State.PLAYER2 : State.PLAYER1;
        let { x: crow, y: ccol } = currentCell;
        let pos = [[-1, -1], [-1, 1], [1, -1], [1, 1]], coords = [];
        for (let [diffRow, diffCol] of pos) {
            let destinationCell = new Cell(crow + 2 * diffRow, ccol + 2 * diffCol);
            let adversaryCell = new Cell(crow + diffRow, ccol + diffCol);
            if (!visitedCells.find(c => c.equals(adversaryCell)) && this.checkers.onBoard(destinationCell) && this.checkers.getState(destinationCell) === State.EMPTY && this.checkers.getState(adversaryCell) === adversaryPlayer) {
                visitedCells.push(adversaryCell);
                let moves = this.possibleMoves(destinationCell, visitedCells);
                if (moves.length === 0) {
                    coords.push(destinationCell);
                } else {
                    moves.forEach(v => v.push(destinationCell));
                    coords.push(moves);
                }
                visitedCells.pop();
            }
        }
        let poss = coords.flat().map(p => p instanceof Cell ? [p] : p);
        let max = Math.max(...poss.map(p => p.length));
        return poss.filter(p => p.length === max);
    }
}